drivers/regulator/regulator_npm6001.c

```diff
@@ -0,0 +1,689 @@
+#define NPM6001_LDO0VOUT_SET3V3	 0x1EU
+
+/* nPM6001 BUCKXCONFPWMMODE fields */
+#define NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM_MSK 0x8U
+#define NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM_POS 3
+#define NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM                                   \
+	BIT(NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM_POS)
+
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE0PADTYPE_CMOS  BIT(0)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE1PADTYPE_CMOS  BIT(1)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE2PADTYPE_CMOS  BIT(2)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE0PULLD_ENABLED BIT(4)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE1PULLD_ENABLED BIT(5)
+#define NPM6001_BUCKMODEPADCONF_BUCKMODE2PULLD_ENABLED BIT(6)
+
+/* nPM6001 PADDRIVESTRENGTH fields */
+#define NPM6001_PADDRIVESTRENGTH_READY_HIGH BIT(2)
+#define NPM6001_PADDRIVESTRENGTH_NINT_HIGH  BIT(3)
+#define NPM6001_PADDRIVESTRENGTH_SDA_HIGH   BIT(5)
+
+struct regulator_npm6001_pconfig {
+	struct i2c_dt_spec i2c;
+	uint8_t buck_pad_val;
+	uint8_t pad_val;
+};
+
+struct regulator_npm6001_config {
+	struct regulator_common_config common;
+	const struct device *p;
+	uint8_t source;
+};
+
+struct regulator_npm6001_data {
+	struct regulator_common_data data;
+};
+
+static const struct linear_range buck0_range =
+	LINEAR_RANGE_INIT(1800000, 100000U, 0x0U, 0xFU);
+
+static const struct linear_range buck1_range =
+	LINEAR_RANGE_INIT(700000, 50000U, 0x0U, 0xEU);
+
+static const struct linear_range buck2_range =
+	LINEAR_RANGE_INIT(1200000, 50000U, 0xAU, 0xEU);
+
+static const struct linear_range buck3_range =
+	LINEAR_RANGE_INIT(500000, 25000U, 0x0U, 0x70U);
+
+static const int32_t ldo0_voltages[] = {
+	1800000, 2100000, 2410000, 2700000, 3000000, 3300000
+};
+
+static int regulator_npm6001_ldo0_list_voltage(const struct device *dev,
+					       unsigned int idx,
+					       int32_t *volt_uv)
+{
+	if (idx >= ARRAY_SIZE(ldo0_voltages)) {
+		return -EINVAL;
+	}
+
+	*volt_uv = ldo0_voltages[idx];
+
+	return 0;
+}
+
+static int regulator_npm6001_buck012_set_voltage(
+	const struct device *dev, int32_t min_uv, int32_t max_uv,
+	const struct linear_range *range, uint8_t vout_reg, uint8_t conf_reg)
+{
+	const struct regulator_npm6001_config *config = dev->config;
+	const struct regulator_npm6001_pconfig *pconfig = config->p->config;
+	uint8_t conf, buf[2];
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_get_win_index(range, min_uv, max_uv, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	buf[0] = (uint8_t)idx;
+	buf[1] = (uint8_t)idx;
+
+	/* force PWM mode when updating voltage */
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, conf_reg, &conf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((conf & NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM) == 0U) {
+		ret = i2c_reg_write_byte_dt(
+			&pconfig->i2c, conf_reg,
+			conf | NPM6001_BUCKXCONFPWMMODE_SETFORCEPWM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* write voltage in both ULP/PWM registers */
+	ret = i2c_burst_write_dt(&pconfig->i2c, vout_reg, buf, sizeof(buf));
```

Comment:
If the above works then this:
```
	buf[0] = vout_reg;
	buf[1] = (uint8_t)idx;
	buf[2] = (uint8_t)idx;

	ret = i2c_write_dt(&pconfig->i2c, vout_reg, buf, sizeof(buf));
```
should work as well. At least on nRF SoCs it should produce the same outcome on the bus (well, for TWIM, only if `zephyr,concat-buf-size` is set to at least 3).

Using `i2c_burst_write_dt()` is risky because of: https://github.com/zephyrproject-rtos/zephyr/blob/8bc094610c5dd69f6ba0085fb3fe7d18167164c9/include/zephyr/drivers/i2c.h#L1220-L1223 https://github.com/zephyrproject-rtos/zephyr/blob/8bc094610c5dd69f6ba0085fb3fe7d18167164c9/include/zephyr/drivers/i2c.h#L175-L182 and for example for the i2c_nrfx_twim driver the result depends on the `zephyr,concat-buf-size` property.

I'd rather restore the separate writes if the combined write done with `i2c_write_dt()` does not work.

Reply:
`i2c_write_dt` works well, updated.